### PR TITLE
[UPDATE] Converted pages to functional components, and aiming to resove issues #17, #18, #19, #20

### LINKED ISSUE #17: Improve code organisation  - Frontend
The code organization is very poor. This reflects bad code quality and hinders the app development. 
Few questions and observations:-
1. Why are tsx files separated into components and pages? Provide logical explanation.
2. Why static files like css and images in two separate folders? Also observed js files with tsx components.
3. Code is not commented at all.
4. Business logic is defined in the same files as state components.

Here is the checklist that should be completed/looked into before closing off this issue:-
- [ ] Organised folder structure
   Inspiration - [Blog link](https://medium.com/swlh/how-to-structure-your-typescript-react-redux-app-877d1eba1c1e)
     code
- [ ] camelCased file names
- [ ] Organised static assets  
- [ ] tsconfig.json should build needed files 
  
Feel free to comment and ask anything.

### PR DESCRIPTION
**Pull request by v-rohan**
**[Changelog]**

_Points marked with a (!) are still a work in progress._

- Improved the file-system structure of the code, as pointed out in #17 (!)
- Implemented Redux in all the pages where it was deemed necessary.
- Fixed the few UI bugs as seen in issue #20 
- Seperated business logic from nearly all pages, as pointed out in #19  components are still left in a few places.
- Furthermore, regarding the defining of types as pointed in #18 , Next JS automatically generates that file in accordance to it's needs, and configuring that file seemed unwise without further introspection. If any changes, are possible or still needed, we're/will definitely work on them.
- The entire code will be commented out as soon as we're successful in abstracting all the business logic and CSS from components/pages(!)

In accordance to the following query,
> Why are tsx files separated into components and pages? Provide logical explanation.

as seen in #17, next.js automatically makes two folders named **pages** and **components**, and in the pages folder, each .ts/.tsx/.js file is automatically given a subdomain name corresponding to the sub-folder/file's name. Thus, JSX components which are re-purposed in a lot of pages are stored in the components folder to avoid code repetition and Next routing errors.

@arc9693 , @himanshu272 Kindly, review and let us know if any further changes are needed.

### CODE REVIEW

**Comment by realsdx** [comment]
I think, you accidentally changed the mode of every files in the project. Push a update to correct that.

**Comment by arc9693** [comment]
@rishav4101 why exactly have you used functional components instead of class components?

**Comment by rishav4101** [comment]
1) Class component is getting deprecated slowly after the introduction of react Hooks, and so is the support of class components.

2) Redux in nextjs uses next-redux-wrapper which cannot be used with class components. 

3) Class component is less reactive than functional components.

**Review by arc9693 — COMMENTED** [review]
@rishav4101 
<ol>
<li> A few lines from the react docs:
  
> TLDR: There are no plans to remove classes from React.

      [LINK](https://reactjs.org/docs/hooks-intro.html#gradual-adoption-strategy)

  > We don’t recommend rewriting your existing classes to Hooks unless you planned to rewrite them anyway (e.g. to fix bugs).

  > It is an early time for Hooks, and some third-party libraries might not be compatible with Hooks at the moment.</li>
<li>To my understanding, we are using next.js for building static files for SSR. There are multiple alternatives to using next.js that can be easily integrated in our project without "migration of all class components to react hooks". 
</li>
<li>
Functional components are simpler. React hooks help you do what you could already do with class components, with more freedom and less complexity. It is indeed the future. But a migration like this should be discussed before implementation. There are a number of aspects to be looked into.This needs more discussion. I would not recommend it right away. 

@realsdx What are your views?
</li>
</ol>

**Comment by realsdx** [comment]
We are not doing next.js for sure. As I already discussed about it with @romitkarmakar and @himanshu272 . Our app has very few components , it should not require SSR at all.

And about the Functional component, this type of things should be discussed before.

Anyways, if you guys think we should shift to functional component , then let's discuss it's pros and cons, then we can take a decision on it.

**Comment by arc9693** [comment]
@v-rohan @rishav4101 
I recommend working on each issue separately. 
SSR is not our focus right now. If you guys want to work on migration, open up a new issue. Write your pros and cons. Let's discuss it first.

File: website/pages/question/description2.tsx [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+      setTimeout(() => this.setState({ copied: false }), 1500);
+    });
+  }
+  submitcode = (code: any, lang: any) => {
+    this.setState({
+      isLoading: true,
+      res: [],
+    });
+    var self = this;
+    fetch(`${process.env.NEXT_PUBLIC_BACKEND_URL}/api/submit`, {
```
[the remark this answers is not in the document]
Comment (v-rohan) [reply website/pages/question/description2.tsx:146]: This was an auxiliary page based on the class components i.e, the original pages/description.tsx I was still making changes to this page, while making the PR. This page will be removed soon.

**Comment by himanshu272** [comment]
In my opinion it is preferable to convert to functional components since we plan on using Redux and having them all in a single source will help. Class components don't integrate well with it.

**Comment by romitkarmakar** [comment]
@realsdx we have choose Nextjs to create a universal react app. As single page app can't fulfill the application needs of dynamic link generation like unique shareable link for every contest which can be posted on social media and better SEO optimization which is more advanced in universal apps than single page apps. Functional components have been decided instead of class components as HOC are slowly been deprecated for redux.

**Comment by realsdx** [comment]
@romitkarmakar Those two reasons are not enough for the introduction of Next.js , as we don't need them, they are just 'Nice to have'. Also, SSR adds unnecessary complexity to the project. 
Keep it simple(in development, build and deployment). 
And SEO shouldn't be a problem.

File: website/components/newAppbar.tsx [hunk 2]
```diff
@@ -272,17 +280,17 @@ class Newappbar extends Component {
...
     
       <AppBar position="static" elevation={0}>
         
-        <Toolbar disableElevation>
+        <Toolbar>
           <Typography
             variant="h6"
-            className={classes.title}
+            className={this.props.classes.title}
             onClick={() => Router.push("/")}
```
Comment (arc9693) [inline website/components/newAppbar.tsx:294]: Can we have one wrapper function for all the route change calls?

File: website/pages/login/index.tsx [hunk 3]
```diff
@@ -12,7 +11,7 @@ import IconButton from "@material-ui/core/IconButton";
 import InputAdornment from '@material-ui/core/InputAdornment';
 import Visibility from '@material-ui/icons/Visibility';
 import VisibilityOff from '@material-ui/icons/VisibilityOff';
-
+import Router from "next/router";
```
Comment (arc9693) [inline website/pages/login/index.tsx:14]: Why not react-router?

File: website/pages/signup/facebook.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            headers: {
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify({
+                code: getParameterByName("code"),
+            }),
+        })
+            .then((resp) => resp.json())
+            .then((res) => {
+                fetch(`${process.env.NEXT_PUBLIC_BACKEND_URL}/login`, {
```
Comment (arc9693) [inline website/pages/signup/facebook.tsx:39]: Change needed.

**Review by arc9693 — CHANGES_REQUESTED** [review]
Tasks 

- [ ]  Renaming of following files: Timer.tsx, UserContextProvider.tsx

- [ ] Following piece of code and similar, repeated over multiple files should be defined in enums or a constants file
`const dict = {
    c: "c_cpp",
    "c++": "c_cpp",
    python3: "python",
    java: "java",
  };`

- [ ] Move all the API call definitions to one file. Write modular code. Avoid following type of code:

`  onClick={() => {
                      fetch(
                        `${process.env.NEXT_PUBLIC_BACKEND_URL}/account/logout`,
                        {
                          method: "POST",
                          headers: {
                            Authorization: `Token ${localStorage.token}`,
                          },
                        }
                      ).then(() => {
                        localStorage.clear();
                        Router.push("/");
                      });
                    }}`

     Source: Line 238 in newAppBar.tsx

- [ ] Correct the non-uniformity of filenames inside the store/actions and store/reducers. 
     ` eg: contestReducers.tsx and editorialReducer.tsx`
- [ ] Create one public folder and move all the static files to it under sub-folders images, js ,css etc accordingly.
- [ ] Create one utils folder. Move all the helper, service (API calls), constants etc. to it.

**Comment by arc9693** [comment]
@himanshu272  Redux can integrate with class components very well. 
@romitkarmakar If you are using next.js just to create a shareable link, you have got no problem. You don't need next.js for that. This is a very simple project. You don't need HOC in this project in the first place.

**Comment by realsdx** [comment]
`const dict = { c: "c_cpp", "c++": "c_cpp", python3: "python", java: "java", };` frontend should not have this line. Language data should be fetched from server, and cached during initial load.

**Comment by arc9693** [comment]
@v-rohan are all the changes complete on this PR? Kindly fill the checks and mention me, when they are complete.

**Comment by v-rohan** [comment]
@arc9693 Kindly review.

**Comment by arc9693** [comment]
@v-rohan I am putting it under review. You can expect feedback from me by tomorrow EOD.

**Review by arc9693 — APPROVED** [review]
@v-rohan Most of the code has been organized. Except for some issues that we can take up separately, this PR looks good to be merged now. Good work.

**Comment by v-rohan** [comment]
Hello @arc9693, @realsdx, @himanshu272, @romitkarmakar 

It is a bit unprofessional, but this PR is the thing I'll remember the most from all my time in GLUG. I was reminiscing about this time in our year group, now that we have received our farewell recently. This gave me so much trouble at the time, but whenever I'd see juniors slacking off and thinking about accepting it, I'd remember the standards you guys set for us. Thanks.

**Comment by realsdx** [comment]
@v-rohan I am a proud alumni of GLUG, thanks to all of you.

**Comment by arc9693** [comment]
This just made my day.
We are all very proud of you guys.
Keep up the good work.
Best of luck for your future endeavours. ❤️
